### PR TITLE
LetDirective template binding - tests, docs and demo

### PR DESCRIPTION
**Pull request by kajetansw**
(This PR is a direct continuation of this one: https://github.com/BioPhoton/rx-angular/pull/178)

## Description

This PR adds `ng-template` bindings to be used with `LetDirective`. With this, we can create and bind templates for different state of an observable:
- **next** - when values are emitted
- **complete** - on observable completion 
- **error** - on observable error
- 🔥 **suspense** 🔥 - render something before the first value is emitted from an observable

## Checklist

Checklist of things to be done for the feature:
- [x] Implement the feature itself
- [x] Ensure that all previous test are passing and the previous functionality still works
- [x] Add tests
- [x] Refactor the `TemplateManager` class to a factory function
- [x] Update docs for LetDirective (jsdocs and Markdown)
- [x] Add comments explaining low-level implementation details
- [x] Add new page with demo in the `template-demo` app

## Example

Here is some small, example component that uses the newest syntax:

```ts
@Component({
  selector: 'let-template-binding',
  template: `
    <button [unpatch] (click)="signals$?.complete()">
      complete
    </button>
    <button [unpatch] (click)="signals$?.next(random())">
      next
    </button>
    <button [unpatch] (click)="signals$?.error(errorStub)">
      error
    </button>

    <div
      *rxLet="
        signals$;
        let count;
        strategy: visibleStrategy;
        complete: complete;
        error: error;
        suspense: suspense
      "
    >
      value: {{ count | json }}
    </div>

    <ng-template #complete>
      <h1>COMPLETE</h1>
    </ng-template>
    <ng-template #error>
      <h1>ERROR</h1>
    </ng-template>
    <ng-template #suspense>
      <h1>SUSPENSE</h1>
    </ng-template>
  `,
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class LetTemplateBindingComponent {
  errorStub = new Error("Test error");

  signals$ = new Subject<any>();
  // signals$ = fromPromise(fetch('https://swapi.dev/api/people/1').then(a => a.json()));
  // signals$ = from([1,2,3,4,5]).pipe(first());

  // signals$ = interval(1000).pipe(
  //   switchMap(() => fromPromise(fetch(`https://swapi.dev/api/people/${Math.floor(Math.random() * 20)}`).then(a => a.json()))),
  //   takeUntil(interval(5000))
  // );

  visibleStrategy = 'local';

  random() {
    return Math.random();
  }
}
```

## Demo

Inside `template-demo` app there's a new page - "Template Binding". There you can play with two example scenarios with usage of `*rxLet` and template binding:
- manual `Subject` triggering,
- triggering HTTP request for an external API on certain time intervals.

## Any questions?

Comments are welcome! Feature is also ready to play with, so if you want you can switch to my branch and give it a try.

### CODE REVIEW

File: libs/template/docs/let.md [hunk 1]
```diff
@@ -14,6 +14,7 @@ Other Features:
 - it takes away multiple usages of the `async` or `push` pipe
 - a unified/structured way of handling null, undefined or error
 - distinct same values in a row skip not needed re-renderings
+- display custom templates for different observable notifications
```
Comment (BioPhoton) [inline libs/template/docs/let.md:17]:
```suggestion
- display custom templates for different observable notifications (suspense, next, error, complete)
```

File: libs/template/spec/let/let.directive.template-binding.all.spec.ts [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+  });
+
+  it('should render "suspense" template before the first value is emitted', () => {
+    component.value$ = new Subject();
+    fixture.detectChanges();
+    expectContentToBe('suspense');
+  });
+
+  it('should render "complete" template on the observable completion', () => {
+    fixture.detectChanges();
```
Comment (BioPhoton) [inline libs/template/spec/let/let.directive.template-binding.all.spec.ts:57]: I would make it more clear why it completes.
```suggestion
    component.value$ = EMPTY;
    fixture.detectChanges();
```

File: libs/template/spec/let/let.directive.template-binding.all.spec.ts [hunk 3]
```diff
@@ -0,0 +1,104 @@
+import { Component } from '@angular/core';
+import { async, ComponentFixture, fakeAsync, TestBed, tick } from '@angular/core/testing';
+import { LetDirective } from '@rx-angular/template';
+import { EMPTY, interval, Observable, of, Subject, NEVER } from 'rxjs';
+import { take } from 'rxjs/operators';
```
Comment (BioPhoton) [inline libs/template/spec/let/let.directive.template-binding.all.spec.ts:5]:
```suggestion
import { take, throwError } from 'rxjs/operators';
```

File: libs/template/spec/let/let.directive.template-binding.all.spec.ts [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+  });
+
+  it('should render "error" template on observable error', () => {
+    component.value$ = new Subject();
+    fixture.detectChanges();
+
+    (component.value$ as Subject<number>).error(new Error('test'));
+    fixture.detectChanges();
+
+    expectContentToBe('error');
```
Comment (BioPhoton) [inline libs/template/spec/let/let.directive.template-binding.all.spec.ts:68]: I suggest to use throwError here
```suggestion
    component.value$ = throwError('Error');
    fixture.detectChanges();
    expectContentToBe('error');
```

File: libs/template/spec/let/let.directive.template-binding.all.spec.ts [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+    fixture.detectChanges();
+    expectContentToBe('0');
+
+    tick(1000);
+    fixture.detectChanges();
+    expectContentToBe('1');
+
+    tick(1000);
+    fixture.detectChanges();
+    expectContentToBe('complete');
```
Comment (BioPhoton) [inline libs/template/spec/let/let.directive.template-binding.all.spec.ts:86]:
```suggestion
    // As the last value ('2') and the compelete notification are sync we expect complete here 
    expectContentToBe('complete');
```

File: libs/template/spec/let/let.directive.template-binding.no-complete.spec.ts [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+
+@Component({
+  template: `
+    <ng-container *rxLet="value$; let value; suspense: suspense; error: error">{{
+      value === undefined
+        ? 'undefined'
+        : value === null
+        ? 'null'
+        : (value | json)
+    }}</ng-container>
```
Comment (BioPhoton) [inline libs/template/spec/let/let.directive.template-binding.no-complete.spec.ts:14]: Maybe its useful to use the `let e = error; let c = complete;` ?

File: libs/template/src/lib/core/utils/template-manager_creator.ts [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+ * TemplateManager
+ *
+ * @description
+ * This function returns an object that holds the logic for managing templates to
+ * be embedded into a `ViewContainerRef`. Internally it caches template references and instantiates
+ * embedded views. The `TemplateManager` lets you re-use templates and insert views on-demand (e.g.
+ * when a new observable notification is sent) and update the view context.
+ *
+ * @param viewContainerRef reference to a top-level view container templates will be attached to
+ * @param initialViewContext initial view context state
```
Comment (BioPhoton) [inline libs/template/src/lib/core/utils/template-manager_creator.ts:22]:
```suggestion
 * This function returns an object that holds the logic for managing templates of a `ViewContainerRef`. 
 * It abstracts `EmbeddedView` creation, `insert` calls and `ViewContext` updates.
 * Internally the creations of template references are lazy, and a combination of caching logic and the `ViewContainerRef#detach` method.
 * The `TemplateManager` lets you re-use templates and insert views on-demand, as well as update the view context.
 * (e.g. when a new observable notification is sent) 
 *
 * @param viewContainerRef reference to a top-level view container where the passed templates will be attached to
 * @param initialViewContext initial view context state
```

File: libs/template/src/lib/let/let.directive.ts [hunk 8]
```diff
@@ -215,7 +234,7 @@ export class LetDirective<U> implements OnInit, OnDestroy {
   }
 
   ngOnDestroy() {
-    this.subscription.unsubscribe();
+    this.subscription?.unsubscribe();
```
Comment (BioPhoton) [inline libs/template/src/lib/let/let.directive.ts:237]: We could also initialize it `  protected subscription: Unsubscribable = new Subscription();`

File: libs/template/docs/let.md (hunk 1 above)
Comment (kajetansw) [reply libs/template/docs/let.md]: Done.

File: libs/template/spec/let/let.directive.template-binding.all.spec.ts (hunk 2 above)
Comment (kajetansw) [reply libs/template/spec/let/let.directive.template-binding.all.spec.ts:59]: Done.

File: libs/template/spec/let/let.directive.template-binding.all.spec.ts (hunk 3 above)
Comment (kajetansw) [reply libs/template/spec/let/let.directive.template-binding.all.spec.ts:5]: Done.

File: libs/template/spec/let/let.directive.template-binding.all.spec.ts (hunk 4 above)
Comment (kajetansw) [reply libs/template/spec/let/let.directive.template-binding.all.spec.ts]: Done.

File: libs/template/spec/let/let.directive.template-binding.all.spec.ts (hunk 5 above)
Comment (kajetansw) [reply libs/template/spec/let/let.directive.template-binding.all.spec.ts:86]: Done.

File: libs/template/spec/let/let.directive.template-binding.no-complete.spec.ts (hunk 6 above)
Comment (kajetansw) [reply libs/template/spec/let/let.directive.template-binding.no-complete.spec.ts:14]: I don't know, in what way would it be useful?

File: libs/template/src/lib/core/utils/template-manager_creator.ts (hunk 7 above)
Comment (kajetansw) [reply libs/template/src/lib/core/utils/template-manager_creator.ts]: Updated the jsdoc and applied the suggestion with some minor tweaks ;)

File: libs/template/src/lib/let/let.directive.ts (hunk 8 above)
Comment (kajetansw) [reply libs/template/src/lib/let/let.directive.ts]: Done.

**Comment by kajetansw** [comment]
@BioPhoton thanks for the suggestions!
I thought about the name and didn't come up with anything good... Out of your suggestions the `LazyTemplateManager` might be the best one.
I just don't want to end up with something straight from Java's Spring like `CacheableAbstractTemplateProxyManagerFactory` :D Nevertheless, I understand, that `TemplateManager` might look too opaque.
Naming things is hard.

File: libs/template/spec/let/let.directive.template-binding.no-complete.spec.ts (hunk 6 above)
Comment (BioPhoton) [reply libs/template/spec/let/let.directive.template-binding.no-complete.spec.ts:14]: In an error case without template to see that the error value is set? 
I guess we have that in another test already. So not that important

File: libs/template/spec/let/let.directive.template-binding.all.spec.ts [hunk 9]
```diff
@@ -56,6 +56,8 @@ describe('LetDirective when template binding with all templates', () => {
   it('should render "complete" template on the observable completion', () => {
     // The resulting synchronous notification sequence is: 1,2,3,complete
     component.value$ = of(1, 2, 3);
+    // The resulting synchronous notification sequence is: 1,2,3,complete 
+    component.value$ = of(1,2,3);
```
Comment (kajetansw) [inline libs/template/spec/let/let.directive.template-binding.all.spec.ts:60]: @BioPhoton I have already added those lines.

Comment (BioPhoton) [reply libs/template/spec/let/let.directive.template-binding.all.spec.ts:60]: uups :D

**Comment by kajetansw** [comment]
🎉🎉🎉

**Comment by BioPhoton** [comment]
1.0.0-alpha.2 Should include the pr
😁

**Pull request merged**